go/vt/vtctl/workflow/server.go

```diff
@@ -73,6 +75,154 @@ func NewServer(ts *topo.Server, tmc tmclient.TabletManagerClient) *Server {
 	}
 }
 
+// GetCellsWithShardReadsSwitched returns the topo cells partitioned into two
+// slices: one with the cells where shard reads have been switched for the given
+// tablet type and one with the cells where shard reads have not been switched
+// for the given tablet type.
+//
+// This function is for use in Reshard, and "switched reads" is defined as if
+// any one of the source shards has the query service disabled in its tablet
+// control record.
+func (s *Server) GetCellsWithShardReadsSwitched(
+	ctx context.Context,
+	keyspace string,
+	si *topo.ShardInfo,
+	ttype topodatapb.TabletType,
+) (cellsSwitched []string, cellsNotSwitched []string, err error) {
+	cells, err := s.ts.GetCellInfoNames(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, cell := range cells {
+		srvks, err := s.ts.GetSrvKeyspace(ctx, cell, keyspace)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Checking one shard is enough.
+		var (
+			shardServedTypes []string
+			found            bool
+			noControls       bool
+		)
+
+		for _, partition := range srvks.GetPartitions() {
+			if ttype != partition.GetServedType() {
+				continue
+			}
+
+			// If reads and writes are both switched it is passible that the
+			// shard is not in the partition table.
+			for _, shardReference := range partition.GetShardReferences() {
+				if key.KeyRangeEqual(shardReference.GetKeyRange(), si.GetKeyRange()) {
+					found = true
+					break
+				}
+			}
+
+			// It is possible that there are no tablet controls if the target
+			// shards are not yet serving, or once reads and writes are both
+			// switched.
+			if len(partition.GetShardTabletControls()) == 0 {
+				noControls = true
```

Comment:
Good catch on the `noControls` thing. Let's check with @sougou. I think there are other problems with the checks here. My understanding is a follows:
* A serving type is migrated iff:
a) The tablet_type / shard if found in the serving graph.
b) There are no tablet controls set for this tablet_type / shard.

First problem I see is that the noControls check seems tautological: Having found=true and noControls, means the shard is migrate.

The second problem I see is the following condition:

```
if !tabletControl.GetQueryServiceDisabled() {
						shardServedTypes = append(shardServedTypes, si.ShardName())
}
```
Isn't this adding to the list shards that actually have QueryService Enabled?


Furthermore, the presence of TabletControls should indicate that a migration is in progress, but ultimately the routing is determined by what is the serving graph. The more I think about this, it seems that we should just remove all these checks.

Reply:
@rohit-nayak-ps  do you have a different understanding here? I'm more familiarized with the legacy MigrateServedTypes and maybe the logic changed.